### ISSUE TITLE
[BUG] BlobClient#upload not releasing file handles

### ISSUE THREAD

**Issue by strangelookingnerd** [Azure/azure-sdk-for-java#35971]
**Describe the bug**
I'm trying to upload a file via `FileInputStream` using the `BlobClient`.
The following example throws a `FileSystemException` concerning the file being used by another process.
The stream is closed by the `try-with-resource` block so that does not seem to be an issue.

***Exception or Stack Trace***
FileSystemException / File can not be deleted due to being used by another process

**To Reproduce**
* Create a FileInputStream from a file
* Upload stream via BlobClient#upload
* Close stream
* Delete the file 
* File can not be deleted

***Code Snippet***

```java
File file = new File("test.txt");
Files.write(file.toPath(), "Test".getBytes(Charset.defaultCharset()));

BlobContainerClient client = new BlobContainerClientBuilder() //
    .connectionString("connectionString") //
    .containerName("containerName") //
    .buildClient();

try (FileInputStream stream = new FileInputStream(file)) {
    client.getBlobClient("blobName").upload(stream, true);
}

Files.delete(file.toPath()); // throws java.nio.file.FileSystemException
```

**Expected behavior**
File can be deleted after the upload is completed and open file handles are released. 

**Screenshots**
-

**Setup:**
 - OS: Windows
 - IDE: IntelliJ and Eclipse
 - Library/Libraries: com.azure:azure-storage-blob:12.22.3
 - Java version: 11 and 17
 - App Server/Environment: Plain Java Application 
 - Frameworks: Plain Java Application 

**Additional context**
Using `BlobClient#uploadFromFile` with the file instead of the `FileInputStream` works fine. However in my scenario I only have a `FileInputStream` that I want to upload.

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

**Comment by alzimmermsft** [comment]
Thank you for filing this issue @strangelookingnerd.

As a potential work around for this issue could you try using `BlobClient.uploadFromFile(String filePath, boolean overwrite)` instead as this uses different logic for accessing and uploading the file.

From your sample:

```java
File file = new File("test.txt");
Files.write(file.toPath(), "Test".getBytes(Charset.defaultCharset()));

BlobContainerClient client = new BlobContainerClientBuilder() //
    .connectionString("connectionString") //
    .containerName("containerName") //
    .buildClient();

client.getBlobClient("blobName").uploadFromFile(file.getAbsolutePath(), true);

Files.delete(file.toPath()); // throws java.nio.file.FileSystemException
```

What I believe may be the issue, when `BlobClient.upload(InputStream, boolean)` is used and the `InputStream` is an instance of `FileInputStream` `MappedByteBuffer`s will be used to reduce the need to load the file into JVM memory. It's possible something with Windows and `MappedByteBuffer` may cause issues where the file handles aren't released properly.

@ibrahimrabab FYI

**Comment by strangelookingnerd** [comment]
> As a potential work around for this issue could you try using `BlobClient.uploadFromFile(String filePath, boolean overwrite)` instead as this uses different logic for accessing and uploading the file.

Yes, that would work. However as I said in my use case I only have the FileInputStream. My current workaround is to wrap that into an ByteArrayInputStream. I narrowed the issue down to be caused by „non-markable“ streams due to a hint mentioned in the JavaDocs. 

> What I believe may be the issue, when `BlobClient.upload(InputStream, boolean)` is used and the `InputStream` is an instance of `FileInputStream` `MappedByteBuffer`s will be used to reduce the need to load the file into JVM memory. It's possible something with Windows and `MappedByteBuffer` may cause issues where the file handles aren't released properly.

That might be true. Yet I‘d at least expect a hint to that behavior in the JavaDocs or even better it to be fixed all together.

**Comment by alzimmermsft** [comment]
This has been fixed in the release of `azure-storage-blob:12.23.1` (https://mvnrepository.com/artifact/com.azure/azure-storage-blob/12.23.1), or any library using `azure-core:1.42.0`.
